### PR TITLE
Akka serializer for the SpannerOffset

### PR DESCRIPTION
**Pull request by patriknw**
* needed for offset storage in Akka Projections

### CODE REVIEW

File: journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+/**
+ * INTERNAL API: Serialized SpannerOffset is used by Akka Projections
+ */
+@InternalApi private[akka] class SpannerSerializer(val system: ExtendedActorSystem)
+    extends SerializerWithStringManifest
+    with BaseSerializer {
+  private val SpannerOffsetManifest = "a"
+
+  // persistenceId and commitTimestamp must not contain this separator char
+  private val separator = ';'
```
Comment (patriknw) [inline journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala:26]: do you think there is a risk that we hit a collision for this char? anyone using it in the persistenceId?

File: journal/src/test/scala/akka/persistence/spanner/internal/SpannerSerializerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import akka.actor.ExtendedActorSystem
+import akka.actor.testkit.typed.scaladsl.LogCapturing
+import akka.actor.testkit.typed.scaladsl.ScalaTestWithActorTestKit
+import akka.persistence.spanner.SpannerOffset
+import akka.serialization.SerializationExtension
+import org.scalatest.wordspec.AnyWordSpecLike
+
+class SpannerSerializerSpec extends ScalaTestWithActorTestKit with AnyWordSpecLike with LogCapturing {
+  private val serializer = new SpannerSerializer(system.classicSystem.asInstanceOf[ExtendedActorSystem])
+  private val commitTimestamp = Timestamp.valueOf("2020-06-24 10:35:15.123456").toString
```
Comment (patriknw) [inline journal/src/test/scala/akka/persistence/spanner/internal/SpannerSerializerSpec.scala:17]: the format of the timestamp doesn't really matter as long as it doesn't contain the separator char, but do you know what it looks like for real?

File: journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      )
+  }
+
+  private def offsetFromBinary(bytes: Array[Byte]): SpannerOffset = {
+    val str = new String(bytes, UTF_8)
+    try {
+      val parts = str.split(separator)
+      if (parts.length == 3) {
+        // optimized for the normal case
+        SpannerOffset(parts(0), Map(parts(1) -> parts(2).toLong))
```
Comment (patriknw) [inline journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala:71]: is my assumption right that the normal case will be 1 pid/seqNr?

File: journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala (hunk 1 above)
Comment (johanandren) [reply journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala:27]: Perhaps a more unusual separator char would be good. I suggest 〷 (telegraphy line separator) ;)

File: journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala (hunk 3 above)
Comment (johanandren) [reply journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala:72]: I think it only ever has seen entries when two writes were at the same timestamp. But I'm not 100% sure I remember right.

File: journal/src/test/scala/akka/persistence/spanner/internal/SpannerSerializerSpec.scala (hunk 2 above)
Comment (johanandren) [reply journal/src/test/scala/akka/persistence/spanner/internal/SpannerSerializerSpec.scala:17]: It's rfc3339, you can use `SpannerUtils.*` to format

File: journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala (hunk 1 above)
Comment (patriknw) [reply journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala:27]: at least `;` is escaped with URLEncoder if someone for strange reason is url-encoding their persistenceIds

File: journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala (hunk 3 above)
Comment (chbatey) [reply journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala:72]: @patriknw is right. There's always at least one, the event found at this timestamp. So when we start from that timestamp again it is used to filter out the event from last time.

Comment (chbatey) [reply journal/src/main/scala/akka/persistence/spanner/internal/SpannerSerializer.scala:72]: And yes it is nearly always only 1, unless multiple were written in the same microsecond

**Comment by patriknw** [comment]
@hseeberger Let us know if you start using Akka Projections with Spanner. Then you need a new version of akka-persistence-spanner, that is including this serializer.

**Comment by hseeberger** [comment]
Thanks, @patriknw. Currently we are planing to use Akka Projections for the management cluster only where we are using Postgres as our database.

**Pull request merged**